state/watcher.go

```diff
@@ -1095,6 +1106,202 @@ func (w *relationUnitsWatcher) loop() (err error) {
 	}
 }
 
+// relationStatusWatcher sends notifications when the life or status
+// of the specified relations change.
+type relationStatusWatcher struct {
+	commonWatcher
+	relationKeys set.Strings
+	knownLife    map[string]Life
+	knownStatus  map[string]status.Status
+	out          chan []corewatcher.RelationStatusChange
+}
+
+// WatchStatus returns a watcher that notifies of changes to the life
+// or status of the relation.
+func (r *Relation) WatchStatus() RelationStatusWatcher {
+	return newRelationStatusWatcher(r.st, r.doc.Key)
+}
+
+func newRelationStatusWatcher(backend modelBackend, keys ...string) RelationStatusWatcher {
+	w := &relationStatusWatcher{
+		commonWatcher: newCommonWatcher(backend),
+		out:           make(chan []corewatcher.RelationStatusChange),
+		relationKeys:  set.NewStrings(keys...),
+		knownLife:     make(map[string]Life),
+		knownStatus:   make(map[string]status.Status),
+	}
+	go func() {
+		defer w.tomb.Done()
+		defer close(w.out)
+		w.tomb.Kill(w.loop())
+	}()
+	return w
+}
+
+// Changes returns a channel that will receive the changes to
+// relation life or status. The first event on the
+// channel holds the initial state of the relations.
+func (w *relationStatusWatcher) Changes() <-chan []corewatcher.RelationStatusChange {
+	return w.out
+}
+
+type lifeStatusDoc struct {
+	Id     string        `bson:"_id"`
+	Life   Life          `bson:"life"`
+	Status status.Status `bson:"status"`
+}
+
+var lifeStatusFields = bson.D{{"_id", 1}, {"life", 1}, {"status", 1}}
+
+func (w *relationStatusWatcher) initial() (map[string]Life, map[string]status.Status, error) {
+	coll, closer := w.db.GetCollection(relationsC)
+	defer closer()
+
+	iter := coll.Find(bson.D{{"_id", bson.D{{"$in", w.relationKeys.Values()}}}}).Select(lifeStatusFields).Iter()
+	var doc lifeStatusDoc
+	for iter.Next(&doc) {
+		key := w.backend.localID(doc.Id)
+		if doc.Life != Dead {
+			w.knownLife[key] = doc.Life
+			w.knownStatus[key] = doc.Status
+		}
+	}
+	return w.knownLife, w.knownStatus, iter.Close()
+}
+
+func (w *relationStatusWatcher) merge(updates map[interface{}]bool) (map[string]Life, map[string]status.Status, error) {
+	coll, closer := w.db.GetCollection(relationsC)
+	defer closer()
+
+	// Separate keys into those thought to exist and those known to be removed.
+	var changed []string
+	latestLife := make(map[string]Life)
+	latestStatus := make(map[string]status.Status)
```

Comment:
I have a feeling the watcher would be somewhat easier to read if you introduced a struct which holds Life and Status. This would be replace the 2 `known*` and 2 `latest*` maps with one each and avoid a lot of doubled up map manipulations.